apollo_fpga/__init__.py

```diff
@@ -376,3 +411,78 @@ def get_usb_api_version(self):
     def get_usb_api_version_string(self):
         (api_major, api_minor) = self.get_usb_api_version()
         return (f"{api_major}.{api_minor}")
+
+    @classmethod
+    def print_info(cls, ids=None, stub_ids=None, force_offline=False, timeout=5000, out=print):
+        """ Print information about Apollo and all connected Apollo devices.
+
+        Return True if any connected device is found.
+        """
+        out(f"Apollo version: {__version__}")
+        out(f"Python version: {sys.version}\n")
+
+        found_device = False
+        if ids is None:
+            ids = cls.APOLLO_USB_IDS
+        if stub_ids is None:
+            stub_ids = cls.LUNA_USB_IDS
+
+        # Look for devices with stub interfaces.
+        stub_devs = cls._find_all_devices(stub_ids, cls._device_has_stub_iface)
+
+        for device in stub_devs:
+            found_device = True
+            out("Found Apollo stub interface!")
+            out(f"\tBitstream: {device.product} ({device.manufacturer})")
+            out(f"\tVendor ID: {device.idVendor:04x}")
+            out(f"\tProduct ID: {device.idProduct:04x}")
+            out(f"\tbcdDevice: {device.bcdDevice:04x}")
+            out(f"\tFlash UID: {device.serial_number}")
```

Comment:
Can we safely assume that the device serial number will always be equal to the flash UID?

Reply:
I think calling it Serial number might be confusing because in the context of Apollo, that usually means the MCU serial number.

And yes, there could be future Apollo-based devices that choose to use something else as the serial number, including devices developed by folks other than GSG.

Reply:
I had some doubt about this myself. Currently it is called "Flash UID" in `cynthion info` but only since some hastily-written code for the 0.1.0 release. This chunk of output only applies to gateware with an Apollo stub interface for a shared USB port, so it won't affect any current platforms other than Cynthion and URTI. If there are future platforms with shared USB that don't have flash or don't use the flash UID for bitstream serial number, we can make this output platform-specific. Users won't see this output for their own custom gateware unless the they make the effort to include `ApolloAdvertiser`.

I agree that calling it "Serial number" would be confusing. On the other hand, users are somewhat likely to see it called "SerialNumber" or "iSerial" in places like `dmesg` or `lsusb`. How about "Bitstream serial number" or "Gateware serial number" or "FPGA serial number"?

Reply:
Of the three I guess I would lean toward "Bitstream serial number" since "Bitstream" is already used above.

Reply:
I'm not keen on calling it the FPGA serial number, because that's actually misleading as to what part it comes from on the board. And I'm not keen on calling it the bitstream serial number either, because there's no guarantee that the bitstream uses that value as its serial number. Whereas if we call it the flash UID, that's definitely going to be accurate, regardless of what any future third-party platforms or bitstreams might do.

Reply:
This code is displaying `iSerialNumber` from the bitstream's device descriptor, so I'm not understanding why you think there is no guarantee that the bitstream uses that value as its serial number. "That value" is designated `iSerialNumber`.

Reply:
Sorry, I was getting mixed up between what's reported for stub devices and the Apollo MCU itself.